src/path.rs

```diff
@@ -0,0 +1,140 @@
+use Renderable;
+use value::Value;
+use context::Context;
+use token::Token;
+use token::Token::*;
+use error::{Error, Result};
+
+#[derive(Debug)]
+pub struct IdentifierPath {
+    value: String,
+    indexes: Vec<Value>,
+}
+
+impl Renderable for IdentifierPath {
+    fn render(&self, context: &mut Context) -> Result<Option<String>> {
+        println!("{:?}", self);
+        let value = context
+            .get_val(&self.value)
+            .ok_or_else(|| Error::Render(format!("{} not found in context", self.value)))?
+            .clone();
+
+        let mut counter = self.indexes.len();
+        let result = self.indexes.iter().fold(Ok(&value), |value, index| {
+            // go through error
+            let value = if let Ok(value) = value {
+                value
+            } else {
+                return value;
+            };
+            counter -= 1;
+            match (value, index) {
+                (&Value::Array(ref value), &Value::Num(ref x)) => {
+                    // at the first condition only is_normal is not enough
+                    // because zero is not counted normal
+                    if (*x != 0f32 && !x.is_normal()) || *x < 0f32 ||
+                       x.round() > (::std::usize::MAX as f32) {
+                        return Error::renderer("bad array index");
+                    }
+                    let idx = x.round() as usize;
+                    let value =
+                        value
+                            .get(idx)
+                            .ok_or_else(|| Error::Render("index out of range".to_string()))?;
+                    Ok(value)
+                }
+                (&Value::Array(_), _) => Error::renderer("bad array index type"),
+                (&Value::Object(ref value), &Value::Str(ref x)) => {
+                    let value =
+                        value
+                            .get(x)
+                            .ok_or_else(|| Error::Render("object element not found".to_string()))?;
+                    Ok(value)
+                }
+                (&Value::Object(_), _) => Error::renderer("bad object index"),
+                (value, _) if counter == 0 => Ok(value),
+                _ => Error::renderer("non-indexable element found while indexable expected"),
+            }
+        });
+
+        match result {
+            Ok(result) => result.render(context),
+            Err(e) => Error::renderer(&format!("rendering error: {}", e)),
```

Comment:
You can use `chain_err` for this

```rust
result.chain_err(|| "Failed to render expression")
```

Reply:
Dropped, `?` is much better choice, the chained error wouldn't have added much value.